libp2p/src/builder.rs

```diff
@@ -0,0 +1,1301 @@
+// TODO: Rename runtime to provider
+// TODO: Should we have a timeout on transport?
+// TODO: Be able to address `SwarmBuilder` configuration methods.
+// TODO: Consider moving with_relay after with_other_transport.
+// TODO: Consider making with_other_transport fallible.
+
```

Comment:
> Can this handle custom authentication algorithms? Do we need to be able to handle that?

I don't think so. You can do so via the `with_other_transport` mechanism.

> Can this handle custom base transports, like https://github.com/umgefahren/libp2p-tor? Should it?

Yes, see `with_other_transport`.

> I think as a first step, we don't need to handle these cases. In the long run, looking at #3271, it would be great if we could make all of the upgrade stuff an implementation detail of this `SwarmBuilder` and only expose the bare `Transport` trait. In other words, pack all usecases into this `SwarmBuilder`.
> 
> Do you think that is a good direction?

> Make the easy things easy and the hard things possible.

My number one goal of the new `SwarmBuilder` is the former, i.e. "easy things easy". I suggest we focus on that in this pull request.

As long as making the "hard things possible" doesn't come at the expense of the easy things, I am fine with the direction you are suggesting above.

With more and more `Transport`'s being upgraded out-of-the-box (e.g. `libp2p-quic`, `libp2p-webrtc`, ...), dedicated upgrade mechanisms (e.g. `Transport::upgrade`) become less relevant.